incubator/hnc/internal/reconcilers/hierarchy_config.go

```diff
@@ -106,6 +106,15 @@ func (r *HierarchyConfigReconciler) reconcile(ctx context.Context, log logr.Logg
 	}
 	origNS := nsInst.DeepCopy()
 
+	// In v0.5.x, HNC uses v1alpha1 and the annotation on any subnamespace is
+	// called "subnamespaceOf". In v1alpha2, the namespace annotation is upgraded
+	// into "subnamespace-of". If the deprecated "subnamespaceOf" annotation still
+	// exists, upgrade API and early exit since updating the subnamespace will
+	// enqueue the singleton again.
+	if oldSOf := nsInst.GetAnnotations()[api.DeprecatedSubnamespaceOf]; oldSOf != "" {
+		return r.upgradeSubnamespaceOfAPI(ctx, log, nsInst, oldSOf)
+	}
+
```

Comment:
Why do an early return here? We already know how to write the instances (line 144), so I'd just put this into r.syncWithForest() or something like that. You just need to modify nsInst and then it will be written automatically.

I'd do something like this:

```
func syncWithForest(... nsInst ...) {
   ...
   upgradeSubnamespaceAnnotation(... nsInst ...)
   ...
}

func upgradeSubnamespaceAnnotation(... nsInst ...) {
  if old annotation doesn't exist {
     return
  }
  update nsInst
}
```

Reply:
done.